packages/@lwc/engine-dom/src/renderer-factory.ts

```diff
@@ -21,366 +26,9 @@ import type { RendererAPI } from '@lwc/engine-core';
-        return element.getBoundingClientRect();
-    }
-
-    function querySelector(element: Element, selectors: string): Element | null {
-        return element.querySelector(selectors);
-    }
-
-    function querySelectorAll(element: Element, selectors: string): NodeList {
-        return element.querySelectorAll(selectors);
-    }
-
-    function getElementsByTagName(element: Element, tagNameOrWildCard: string): HTMLCollection {
-        return element.getElementsByTagName(tagNameOrWildCard);
-    }
-
-    function getElementsByClassName(element: Element, names: string): HTMLCollection {
-        return element.getElementsByClassName(names);
-    }
-
-    function getChildren(element: Element): HTMLCollection {
-        return element.children;
-    }
-
-    function getChildNodes(element: Element): NodeList {
-        return element.childNodes;
-    }
-
-    function getFirstChild(element: Element): Node | null {
-        return element.firstChild;
-    }
-
-    function getFirstElementChild(element: Element): Element | null {
-        return element.firstElementChild;
-    }
-
-    function getLastChild(element: Element): Node | null {
-        return element.lastChild;
-    }
-
-    function getLastElementChild(element: Element): Element | null {
-        return element.lastElementChild;
-    }
-
-    function isConnected(node: Node): boolean {
-        return node.isConnected;
-    }
-
-    function assertInstanceOfHTMLElement(elm: any, msg: string) {
-        assertInvariant(elm instanceof HTMLElement, msg);
-    }
-
-    const HTMLElementExported = HTMLElementConstructor as typeof HTMLElement;
-
-    const renderer = {
-        HTMLElementExported,
-        insert,
-        remove,
-        cloneNode,
-        createFragment,
-        createElement,
-        createText,
-        createComment,
-        nextSibling,
-        attachShadow,
-        getProperty,
-        setProperty,
-        setText,
-        getAttribute,
-        setAttribute,
-        removeAttribute,
-        addEventListener,
-        removeEventListener,
-        dispatchEvent,
-        getClassList,
-        setCSSStyleProperty,
-        getBoundingClientRect,
-        querySelector,
-        querySelectorAll,
-        getElementsByTagName,
-        getElementsByClassName,
-        getChildren,
-        getChildNodes,
-        getFirstChild,
-        getFirstElementChild,
-        getLastChild,
-        getLastElementChild,
-        isConnected,
-        assertInstanceOfHTMLElement,
-        defineCustomElement,
-        getCustomElement,
-    };
+export function rendererFactory(baseRenderer: RendererAPI | null): SandboxableRendererAPI {
+    const renderer = process.env.RENDERER;
     // Meant to inherit any properties passed via the base renderer as the argument to the factory.
     Object.setPrototypeOf(renderer, baseRenderer);
-    return renderer;
+    return renderer as unknown as RendererAPI;
```

Comment:
The rendererFactory return a value dependent on the parameter passed. This works well.

```js
type RendererAPIType<Type> = Type extends RendererAPI ? RendererAPI : SandboxableRendererAPI;

export function rendererFactory<T extends RendererAPI | null>(baseRenderer: T): RendererAPIType<T> {
    ....
    return renderer as unknown as RendererAPIType<T>;
}
```

[Conditional Types](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html)

Reply:
Done. Thanks a lot!